### ISSUE TITLE
Qualified rebindable syntax

### ISSUE THREAD

**Issue by pkamenarsky** [purescript/purescript#3245]
Currently, it's not possible to use `bind`s from multiple modules without explicitly rebinding `bind` in every do block, i.e.

```purescript
import Prelude (bind, pure)
import O as O

bindPrelude = do
  a <- [1, 2, 3]
  pure a

bindO = do
  a <- O.elems O.users
  O.pure a
  where
    bind = O.bind
```

I am proposing the following syntax for the latter use case:

```purescript
bindO = O.do
  a <- O.elems O.users
  O.pure a
```

which would desugar into `O.bind` calls instead of just using the `bind` which is currently in scope. Also, this could potentially be extended to include other keywords like `if` (if https://github.com/purescript/purescript/issues/3164 gets implemented), for example:

```purescript
O.if a
  then b
  else c
```

**Comment by hdgarrood** [comment]
I’m not keen on this, personally. The way it works now is easier to explain and simpler implementation-wise. Also, I think writing `where bind = I.bind` is not particularly onerous, so adding this only seems to me to make a fairly minor difference; for me, it’s not really enough to justify new syntax. Especially because I think it’s quite rare to want to use more than one `bind` in a given module.

**Comment by pkamenarsky** [comment]
It'd be pretty nice to being able to mix-and-match different `binds` forming different DSLs in a single function, i.e:

```
foo :: String
foo = Monoid.do
  "hello"
  show do
    [unit, unit]
    a <- [1, 2, 3]
    b <- [4, 5, 6]
    pure (a + b)
  "world"
```

(example posted by @natefaubion in the Slack channel).

What are the downsides to adding additional syntax? Seems like a pretty trivial change to me.

**Comment by garyb** [comment]
> I’m not keen on this, personally. <snip>

Interesting, I actually much prefer this to what we have! It perhaps looks a little strange to qualify a keyword until you know what it does, sure, but I would find this much more convenient to use when using indexed monads and the like (and especially when mixing indexed and non-indexed).

**Comment by hdgarrood** [comment]
I think any syntax change is a big deal, because it can affect a really large proportion of the ecosystem — potentially lots of the compiler as well as tooling — and because it’s hard to fully predict the knock-on effects it can have. To give an example, this change would break a source code formatter, for instance. Just because the initial implementation is easy to do that doesn’t mean it won’t have a higher long term maintenance cost.

It also means we have more documentation work to do, and that it becomes harder for people to understand everything in the language reference (because there’s more of it to digest).

More generally, I am a little concerned by the rate at which we are being tempted to add new features (see https://purescript-users.ml/t/ps-future-new-features-vs-documentation-spec/167/5).

**Comment by paf31** [comment]
I strongly agree with @hdgarrood. This means every tool needs to support new syntax for no new real functionality.

**Comment by garyb** [comment]
How would you interleave `do`s using different binds without it? Maybe `let bind = IxM.bind in do ...`, but then you'd get complained at by the shadowed name warning I imagine.

**Comment by paf31** [comment]
```
with :: a -> (a -> b) -> b
with x f = f x

do blah
   x <- with IxM.bind \bind -> do
     ...
   y <- with Prelude.bind \bind -> do
     ...
```

I'm just saying that I consider this at worst a mild inconvenience and not every mild inconvenience deserves its own compiler support.

**Comment by hdgarrood** [comment]
> How would you interleave dos using different binds without it? Maybe let bind = IxM.bind in do ..., but then you'd get complained at by the shadowed name warning I imagine.

I haven’t tried playing around with this so I’m not sure, but if there is no good way of doing this without triggering the shadowed name warning, then I would prefer to address this by removing the shadowed name warning, since there’s an appetite for this anyway and since it leaves us with a simpler language and implementation.

**Comment by hdgarrood** [comment]
What if neither of the `bind` functions you’re using has anything to do with `Monad`?

**Comment by paf31** [comment]
Well they have to be compatible if you're using them in some common outer do block, but otherwise I don't think it matters.

**Comment by JordanMartinez** [comment]
Shouldn't this be closed as of `0.12.2`?
